substrate/frame/staking/src/pallet/impls.rs

```diff
@@ -981,43 +995,88 @@ impl<T: Config> Pallet<T> {
 	///
 	/// If the nominator already exists, their nominations will be updated.
 	///
-	/// NOTE: you must ALWAYS use this function to add nominator or update their targets. Any access
-	/// to `Nominators` or `VoterList` outside of this function is almost certainly
-	/// wrong.
+	/// NOTE: you must ALWAYS use this function to add nominator or update their targets. Any
+	/// access to `Nominators` or `VoterList` outside of this function is almost certainly wrong.
 	pub fn do_add_nominator(who: &T::AccountId, nominations: Nominations<T>) {
-		if !Nominators::<T>::contains_key(who) {
-			// maybe update sorted list.
-			let _ = T::VoterList::on_insert(who.clone(), Self::weight_of(who))
-				.defensive_unwrap_or_default();
-		}
-		Nominators::<T>::insert(who, nominations);
+		match (Nominators::<T>::contains_key(who), T::VoterList::contains(who)) {
```

Comment:
I am looking at `EventListeners::on_nominator_add` and `EventListeners::on_nominator_update` and it seems code would be lot more simple and readable if we just had `on_nominator_update(who, prev_nomination, new_nominations)` both at the caller and at the listener impl.. I still feel EventListener trait is bloated more than needed..